datastore/longhorn.go

```diff
@@ -437,3 +439,61 @@ func (s *DataStore) ListEngineImages() (map[string]*longhorn.EngineImage, error)
 	}
 	return itemMap, nil
 }
+
+func (s *DataStore) CreateNode(node *longhorn.Node) (*longhorn.Node, error) {
+	return s.lhClient.LonghornV1alpha1().Nodes(s.namespace).Create(node)
+}
+
+// CreateDefaultNode will set default directory to node replica mount path
+func (s *DataStore) CreateDefaultNode(name string) (*longhorn.Node, error) {
+	disks := map[string]types.Disk{}
+	defaultDisk := types.Disk{
+		Path: longhornDirectory,
+	}
+	disks[longhornDirectory] = defaultDisk
+
+	node := &longhorn.Node{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: name,
+		},
+		Spec: types.NodeSpec{
+			Name:            name,
+			AllowScheduling: true,
+			Disks:           disks,
+		},
+	}
+	return s.CreateNode(node)
+}
+
+func (s *DataStore) GetNode(name string) (*longhorn.Node, error) {
+	result, err := s.nLister.Nodes(s.namespace).Get(name)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil, nil
+		}
+		return nil, err
+	}
+	return result.DeepCopy(), nil
+}
+
+func (s *DataStore) UpdateNode(node *longhorn.Node) (*longhorn.Node, error) {
+	return s.lhClient.LonghornV1alpha1().Nodes(s.namespace).Update(node)
+}
+
+func (s *DataStore) GetNodeList(nodeNameList []string) ([]*longhorn.Node, error) {
+	nodeList := make([]*longhorn.Node, 0)
+	for _, nodeName := range nodeNameList {
+		node, err := s.GetNode(nodeName)
+		if err != nil {
+			return nil, err
+		}
+		if node == nil {
+			node, err = s.CreateDefaultNode(nodeName)
```

Comment:
It's weird to have `GetNodeList` to have a side-effect to create the nonexisting nodes. You can have a manager to create its own node if it doesn't exist when starting up at e.g. https://github.com/rancher/longhorn-manager/blob/master/app/daemon.go#L100 .